vnext/Microsoft.ReactNative.Managed/ReactPromise.cs

```diff
@@ -16,102 +16,143 @@
 
 namespace Microsoft.ReactNative.Managed
 {
-    // Implementation of IReactPromise that represents a JavaScript Promise
-    // which can be passed to the native module as a method parameter.
-    //
-    // Methods with ReactMethodAttribute that use an IReactPromise as the last parameter
-    // will be marked as "promise" and will return a promise when invoked from JavaScript.
-    public class ReactPromise<T> : IReactPromise<T>
+  // Implementation of IReactPromise that represents a JavaScript Promise
+  // which can be passed to the native module as a method parameter.
+  //
+  // Methods with ReactMethodAttribute that use an IReactPromise as the last parameter
+  // will be marked as "promise" and will return a promise when invoked from JavaScript.
+  public class ReactPromise<T> : IReactPromise<T>
+  {
+    private IJSValueWriter m_writer;
+    private MethodResultCallback m_resolve;
+    private MethodResultCallback m_reject;
+
+    public ReactPromise(IJSValueWriter writer, MethodResultCallback resolve, MethodResultCallback reject)
+    {
+      m_writer = writer;
+      m_resolve = resolve;
+      m_reject = reject;
+    }
+
+    //  Successfully resolve the ReactPromise with an optional value.
+    public void Resolve(T value)
     {
-        private IJSValueWriter m_writer;
-        private MethodResultCallback m_resolve;
-        private MethodResultCallback m_reject;
+      // Copy to local variables to run resolve outside of lock
+      MethodResultCallback resolve = null;
+      IJSValueWriter writer = null;
 
-        public ReactPromise(IJSValueWriter writer, MethodResultCallback resolve, MethodResultCallback reject)
+      lock (this)
```

Comment:
>lock (this) [](http://example.com/codeflow?start=6&length=11)

C# guidelines is to never lock on this.
Technically anyone code can take a lock on that object and cause DOS...

Guidance is to add a member to the class
```cs
private object m_lock = new Object();
```
and lock on that.

Reply:
similar for other `lock(this)` in this PR.

Reply:
Good point! Now I remember I used to follow these guidelines. I will change it.